File: algorithms-and-data-structures/algorithms-and-data-structures-2/data-structures/flexible-data-structures/list/list.java
class Celula{
	int valor;
	Celula* prox;
	public Celula(int x){
		valor=x;
		prox=null;	
	}
}
class List{
	Celula* p;
	Celula* u;
	int q;
	public List(){
		Celula c=new Celula(-1);
		p=c;
		u=c;
		q=0;
	}
	void inserirInicio(int x){
		Celula c=new Celula(x);	
		if(q==0){
			p->prox=c;
			u=c;
			q++;	
		}
		else{
			c->prox=p->prox;
			p->prox=c;
			q++;
		}
	}
	void inserirFim(int x){
		Celula c=new Celula(x);
		if(q==0){
			p->prox=c;
			u=c;
			q++;	
		}
		else{
			u->prox=c;
			u=c;
			q++;
		}
	}
	void inserirPos(int x,int pos){
		if(pos>=0 && pos<=q){	
			Celula c=new Celula(x);
			Celula tmp=p;
			int i=0;
			while(i<q && i<pos-1){
				tmp=tmp->prox;
				i++;
			}
			c->prox=tmp->prox;
			tmp->prox=c;
			q++;
			if(pos==q){
				u=c;	
			}
		}	
	}
	Celula removerInicio(){
		Celula resp=p->prox;
		p=resp->prox;
		if(q==1){
			u=p;
		}
		q--;
		return resp;
	}
	Celula removerFim(){
		Celula resp=u;
		Celula tmp=p.prox;
		int i=0;
		while(i<q-1){
			tmp=tmp.prox;
			i++;	
		}
		tmp->prox=resp->prox;
		q--;
		return resp;
	}
	Celula removerPos(int pos){
		if(pos>=0 && pos<q){
			Celula tmp=p->prox;
			int i=0;
			while(i<pos-1){
				tmp=tmp->prox;
				i++;
			}
			Celula resp=tmp->prox;
			tmp.prox=resp->prox;
			return resp;	
		}	
	}
}

File: algorithms-and-data-structures/algorithms-and-data-structures-2/data-structures/flexible-data-structures/list/list.c
typedef struct{
	int valor;
	struct Celula* prox;
}Celula;
Celula* construtorCelula(int x){
	Celula* c=malloc(sizeof(Celula));
	(*c).valor=x;
	(*c).prox=NULL;
	return c;	
}
typedef struct{
	Celula* p;
	Celula* u;
	int q;
}Lista;
Lista* construtorLista(){
	Lista* l=malloc(sizeof(Lista));
	Celula* c=construtorCelula(-1);
	(*l).p=c;
	(*l).u=c;
	(*l).q=0;
	return l;
}
void inserirInicio(Lista* l,int x){
	Celula* c=construtorCelula(x);
	if((*l).q==0){
		c->prox=(*l).u->prox;
		(*l).u->prox=c;
		(*l).u=c;
		(*l).q++;
	}
	else{
		c->prox=(*l).p->prox;
		(*l).p->prox=c;
		(*l).q++;	
	}
	
}
void inserirFim(Lista* l,int x){		
	Celula* c=construtorCelula(x);
	(*l).u->prox=c;
	(*l).u=c;
	(*l).q++;
}
void inserirPos(Lista* l,int x,int pos){
		if(pos>=0 && pos<=l->q){	
			Celula* c=construtorCelula(x);
			Celula* tmp=(*l).p;
			int i=0;
			while(i<pos-1){
				tmp=tmp->prox;
				i++;
			}
			c->prox=tmp->prox;
			tmp->prox=c;
			(*l).q++;
			if(pos==(*l).q){
				l->u=c;
			}	
		}
}
Celula* removerInicio(Lista* l){
	Celula resp=(*l).p->prox;
	(*l).p=resp->prox;
	if((*l).q==1){
		(*l).u=(*l).p;
	}
	(*l).q--;
	return resp;
}
Celula* removerFim(Lista* l){
	Celula* c=(*l)->u;	
	Celula* tmp=l->prox;
	int i=0;
	while(i<(*l).q-1){
		tmp=tmp->prox;	
		i++;
	}
	tmp->prox=NULL;
	(*l).q--;
	return c;
}
Celula* removerPos(Lista* l,int pos){
	if(pos>=0 && pos<(*l).q){
		Celula* tmp=(*l).p->prox;
		Celula* resp;
		int i=0;
		while(i<pos-1){
			tmp=tmp->prox;
			i++;
		}
		resp=tmp->prox;
		tmp->prox=resp->prox;
		return resp;
	}	
}
void print(Lista* l){
	Celula* tmp=(*l).p->prox;
	while(tmp!=NULL){
		printf("%d ",tmp->valor);
		tmp=tmp->prox;
	}
	printf("\n");
}

